cmd/crank/beta/render/render.go

```diff
@@ -150,24 +154,63 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
 	// the desired state returned by the last, and each Function may produce
 	// results.
 	for _, fn := range in.Composition.Spec.Pipeline {
-		req := &fnv1beta1.RunFunctionRequest{Observed: o, Desired: d, Context: fctx}
+		// used to store the resources fetched at the previous iteration
+		var extraResources map[string]*fnv1beta1.Resources
+		// used to store the requirements returned at the previous iteration
+		var requirements *fnv1beta1.Requirements
+		var rsp *fnv1beta1.RunFunctionResponse
+
+		for i := int64(0); i < composite.MaxRequirementsIterations; i++ {
+			req := &fnv1beta1.RunFunctionRequest{Observed: o, Desired: d, Context: fctx, ExtraResources: extraResources}
+
+			if fn.Input != nil {
+				in := &structpb.Struct{}
+				if err := in.UnmarshalJSON(fn.Input.Raw); err != nil {
+					return Outputs{}, errors.Wrapf(err, "cannot unmarshal input for Composition pipeline step %q", fn.Step)
+				}
+				req.Input = in
+			}
 
-		if fn.Input != nil {
-			in := &structpb.Struct{}
-			if err := in.UnmarshalJSON(fn.Input.Raw); err != nil {
-				return Outputs{}, errors.Wrapf(err, "cannot unmarshal input for Composition pipeline step %q", fn.Step)
+			conn, ok := conns[fn.FunctionRef.Name]
+			if !ok {
+				return Outputs{}, errors.Errorf("unknown Function %q, referenced by pipeline step %q - does it exist in your Functions file?", fn.FunctionRef.Name, fn.Step)
 			}
-			req.Input = in
-		}
 
-		conn, ok := conns[fn.FunctionRef.Name]
-		if !ok {
-			return Outputs{}, errors.Errorf("unknown Function %q, referenced by pipeline step %q - does it exist in your Functions file?", fn.FunctionRef.Name, fn.Step)
-		}
+			rsp, err = fnv1beta1.NewFunctionRunnerServiceClient(conn).RunFunction(ctx, req)
+			if err != nil {
+				return Outputs{}, errors.Wrapf(err, "cannot run pipeline step %q", fn.Step)
+			}
 
-		rsp, err := fnv1beta1.NewFunctionRunnerServiceClient(conn).RunFunction(ctx, req)
-		if err != nil {
-			return Outputs{}, errors.Wrapf(err, "cannot run pipeline step %q", fn.Step)
+			reqs := rsp.GetRequirements()
+			if reflect.DeepEqual(reqs, requirements) {
+				// the requirements stabilized, the function is done
+				break
+			}
+
+			// if we reached the maximum number of iterations we need to return an error
+			if i == composite.MaxRequirementsIterations-1 {
+				return Outputs{}, errors.Errorf("pipeline step %q reached the maximum number of iterations (%d)", fn.Step, composite.MaxRequirementsIterations)
```

Comment:
This error could be a little clearer as to why we reached maximum iterations - i.e. the requirements didn't stabilize. I'd also personally prefer/expect this "limit exceeded" check at the beginning of the loop.